### PR TITLE
[#44] Migrate to Webpack instead of Browserify

### PR DESCRIPTION
**Pull request by andrew--r**
@iamakulov do I need to add some information to README.md?

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+var webpack = require ('webpack')
```
Comment (vitkarpov) [inline webpack.config.js]: `;`?

@iamakulov Maybe we should to check with eslint gulpfile and now webpack config too?

**Comment by vitkarpov** [comment]
👍

File: webpack.config.js (hunk 1 above)
Comment (iamakulov) [inline webpack.config.js]: @vitkarpov Good idea. #60

File: gulpfile.js [hunk 2]
```diff
@@ -38,7 +36,7 @@ gulp.task ('css', function () {
 
 gulp.task ('zip', ['js', 'css'], function () {
   var version = require ('./package.json').version
-  
+
```
Comment (iamakulov) [inline gulpfile.js]: Could you please revert these formatting changes? We want each commit’s diff to contain only those changes that the commit is related to. We usually ask to move such formatting changes to a separate commit, but as we now have #60 that will do them anyway, it’s OK to just revert them.

Comment (iamakulov) [inline gulpfile.js]: Also related to lines 23 and 56.

File: webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+module.exports = {
+    entry: './source/likely.js',
+    output: {
+        filename: 'likely.js',
+        library: 'likely',
```
Comment (iamakulov) [inline webpack.config.js]: Probably we’d better change this to `ilyabirman-likely` to match the npm module name?

File: webpack.config.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+        libraryTarget: 'umd',
+    },
+    plugins: [
+        new webpack.optimize.OccurrenceOrderPlugin(),
+        new webpack.optimize.DedupePlugin(),
+        new webpack.optimize.UglifyJsPlugin(),
```
Comment (iamakulov) [inline webpack.config.js]: We could optimize this a little bit more by setting the `screw-ie8` option to `true`. (It might be called a bit differently; can’t find the proper name from the phone, sorry.)

**Comment by iamakulov** [comment]
@andrew--r, added a few nits. Apart from that, everything looks great!

File: webpack.config.js (hunk 4 above)
Comment (iamakulov) [inline webpack.config.js]: Found it, it’s called `screw_ie8` and is set like here: https://github.com/SimenB/redux/blob/0b9f9ec2adcf642a1be761763ba3a99b428a3e35/webpack.config.js#L31

**Comment by iamakulov** [comment]
> do I need to add some information to README.md?

I believe no.

File: webpack.config.js (hunk 3 above)
Comment (vitkarpov) [inline webpack.config.js]: @iamakulov Seems, in that case, it breaks [the current behaviour](https://github.com/ilyabirman/Likely/blob/master/source/likely.js#L5). It should be precisely `likely`.

File: source/likely.js [hunk 5]
```diff
@@ -2,5 +2,6 @@
 
 var likely = require('./index.js');
 
-window.likely = likely;
 window.addEventListener('load', likely.initiate);
```
Comment (iamakulov) [inline source/likely.js]: I’m here, sorry for the delay.

I’m very doubtful about this line. If we switch to UMD build for all environments, this will make likely auto-initialize when used [as a CommonJS module](https://github.com/ilyabirman/Likely#using-as-a-commonjs-module) which breaks the current behavior.

We can break the current behavior and release a new major, but I don’t want to do it:
* We can make Likely auto-initialize even when used as a CommonJS module, but it’s a weird thing because a simple `require('ilyabirman-likely')` will cause side effects.
* We can make Likely always require manual initialization, but it’ll make browser users’ life harder which is also bad.

Instead, we should find a way to make Likely auto-initialize itself when being loaded as a separate script. The only reliable solution I can think of is to build two bundles, one for browser and one for CommonJS env, and point `main` in `package.json` to the CommonJS bundle. @vitkarpov, @andrew--r, probably you have any other ideas?

Comment (vitkarpov) [inline source/likely.js]: Maybe we could make it something like this:

```js
(function (root, factory) {
    if (typeof module === 'object' && module.exports) {
        module.exports = factory();
    } else {
        var likely = factory();
        window.addEventListener('load', likely.initiate);
        root.likely = likely;
    }
}(this, function () {
    return likely;
}));
```

?

Comment (iamakulov) [inline source/likely.js]: I’m not sure this will always work. The problem is that regardless of the way someone uses Likely code (by dropping a script into a page, with Webpack, or with Browserify), the script will execute _only_ in browser, where `module` will be undefined. We, however, want to make Likely auto-initialize only when it’s dropped as a `<script>`.

Comment (iamakulov) [inline source/likely.js]: Looks like building two bundles is the only way to go. @andrew--r, do you mind implementing it? If you do mind, just tell me, I’ll do it on my own.

As one of the solutions, we could specify two entry points:
```js
entry: {
  'likely': './source/likely.js',
  'likely-commonjs': './source/index.js'
}
```

which will make Webpack build two different bundles, and update `output.filename` to match the bundle name:

```js
filename: '[name].js'
```

Or feel free to use any other solution if you find it better.

**Comment by iamakulov** [comment]
Could we also update `package.json` so that `main` will target the new release build (instead of the sources)?

File: source/likely.js (hunk 5 above)
Comment (andrew--r) [inline source/likely.js]: I'm afraid I have no free time this week to implement this feature, but I can implement it later, if it's not critical.

Comment (iamakulov) [inline source/likely.js]: Actually, this blocks #48.

Let’s do the following way: take your time and feel free to return to this when it’s OK for you, or I’ll do this on my own earlier if I find some time myself.

Comment (andrew--r) [inline source/likely.js]: Ok :+1:

Comment (vitkarpov) [inline source/likely.js]: Why this doesn't work? If the developer uses Likely as a Commonjs module
than it's the first branch of "if", and we create here Likely instance
without any browser specific behavior. In the other branch we sure that the
bundle wasn't used within Commonjs infrastructure, so we include here ant
additional functionality like auto-init, cause we're sure that the bundle
is used by including it as a separate script in browser. Seems it doesn't
break any compatibility. The only problem I see here is that bundle build
should be configured somehow differs from default. Where I'm wrong?

воскресенье, 22 мая 2016 г. пользователь Andrew Romanov написал:

> In source/likely.js
> <https://github.com/ilyabirman/Likely/pull/58#discussion_r64153025>:
>
> > @@ -2,5 +2,6 @@
> >
> >  var likely = require('./index.js');
> >
> > -window.likely = likely;
> >  window.addEventListener('load', likely.initiate);
>
> Ok 👍
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/ilyabirman/Likely/pull/58/files/2d7471a850b99ed55851d4202bf0f087aa6c6b65#r64153025>
>


-- 
Виктор.

Comment (iamakulov) [inline source/likely.js]: @vitkarpov, look.

Here’s how our Likely bundle looks:

```js
/* likely.js */
(function (root, factory) {
    if (typeof module === 'object' && module.exports) {
        module.exports = factory();
    } else {
        var likely = factory();
        window.addEventListener('load', likely.initiate);
        root.likely = likely;
    }
})
// ...
```

There’s also a user, and they use Webpack in their project. They do `require('likely')` somewhere in their app and then bundle the app. Here’s how their bundle looks now:

```js
/* awesome-app.js */
{
  // ...
  // One of the bundle modules
  42: (function (root, factory) {
      if (typeof module === 'object' && module.exports) {
          module.exports = factory();
      } else {
          var likely = factory();
          window.addEventListener('load', likely.initiate);
          root.likely = likely;
      }
  })
  // ...
}
```

And then the user drops their bundle into a page as a script, and the browser starts executing this:

```js
/* awesome-app.js */
{
  // ...
  // BROWSER: woah, it’s an IIFE, such a lovely IIFE, let’s execute it!
  42: (function (root, factory) {
      // BROWSER: woah, an ‘if’, I love ‘if’s! Let’s see what the condition is...
      // BROWSER: well, as I don’t have ‘module’ in my ‘window’, ‘typeof module’ doesn’t equal ‘object’,
      // BROWSER: so I’ll go to the ‘else’ branch
      if (typeof module === 'object' && module.exports) {
          module.exports = factory();
      } else {
          var likely = factory();
          // BROWSER: cool, an event listener!
          window.addEventListener('load', likely.initiate);
          root.likely = likely;
      }
  })
  // ...
}
```

You see? The `likely.js` code gets to the browser without any changes, so the browsers doesn’t find `window.module` and sets the event listener.

Of course, if a user uses `UglifyJS` plugin, it could (but I believe it wouldn’t) evaluate such `if` in `CommonJS` environment and make that code work the way we want. But this is an assumption about the user’s infrastructure, and we don’t want to make any assumptions.

Have I explained this OK? Or am I wrong in something?

Comment (vitkarpov) [inline source/likely.js]: Seems I get it. My assumption is following: if the developer wants to use Likely as a Commonjs module he's responsible for manual init, seems it's true in many cases. If the developer doesn't want to think about modules and makes a simple website, he just drops the script to the page and Likely works by itself (like now).

Comment (vitkarpov) [inline source/likely.js]: I'm not against two bundles, btw :) It's a good idea.

Comment (iamakulov) [inline source/likely.js]: > My assumption is following: if the developer wants to use Likely as a Commonjs module he's responsible for manual init, seems it's true in many cases. If the developer doesn't want to think about modules and makes a simple website, he just drops the script to the page and Likely works by itself (like now). 

Yep, and this is how it works now. The problem with that IIFE approach is that the event listener is set even if Likely is used as a CommonJS module.

> I'm not against two bundles, btw :)

Yup, I see this :–)

Comment (vitkarpov) [inline source/likely.js]: > The problem with that IIFE approach is that the event listener is set even if Likely is used as a CommonJS module.

Yep. I get it now. Module exists only within bundling runtime and not after that in browser runtime.

Comment (vitkarpov) [inline source/likely.js]: > take your time and feel free to return to this when it’s OK for you, or I’ll do this on my own earlier if I find some time myself.

@andrew--r @iamakulov Guys! Does anybody come back to the task? If everyone's busy I can take it.

Comment (iamakulov) [inline source/likely.js]: @vitkarpov I’m busy till the weekend (at least).

Comment (vitkarpov) [inline source/likely.js]: @andrew--r What about you? I have some time to implement build in two bundles, as we discussed before. You and @iamakulov could just review this on weekends. If you'd like to make it by yourself, than it's okay and I woun't.

**Comment by iamakulov** [comment]
Done some changes and merged. @andrew--r and @vitkarpov, big thanks! 

* Switched Webpack to build two different bundles
* Updated `package.json` to point to `likely-commonjs.js`.
BTW, this `js.js` part is weird, so if you have any better proposal for the name feel free to suggest. `likely-common.js` isn’t OK because `common` is ambiguous.
* Updated the release files. This breaks out current practice of updating them only before releasing a new version, however, this update is required. In this PR, we update the build process and make package.json point to a new file, and if someone installed Likely from GitHub with npm, not updating the release files will break their app.
